compiler/resolution/functionResolution.cpp

```diff
@@ -4286,18 +4301,41 @@ static void generateUnresolvedMsg(CallInfo& info, Vec<FnSymbol*>& visibleFns) {
     int nPrintDetails = 1;
     int nPrint = 3;
 
+
+    // If there is more than one visible function and we haven't been
+    // asked to print all candidates, let's try to eliminate "worse"
+    // matches (e.g., method call / standalone function call
+    // mismatches).
+    Vec<FnSymbol*> filteredFns;
+    if (visibleFns.n > 1 && fPrintAllCandidates == false) {
+      forv_Vec(FnSymbol, fn, visibleFns) {
+        if (!obviousMisMatch(call, fn)) {
+          filteredFns.add(fn);
+        }
+      }
+    }
+
+    // If the filtered functions list is empty (because we skipped the
+    // previous conditional or everything was an obvious mismatch),
+    // then we'll use all the visible functions when printing
+    // candidates in order to avoid suggesting that there are no other
+    // candidates.
+    if (filteredFns.n == 0) {
+      filteredFns = visibleFns;
+    }
```

Comment:
I think it would make the below code simpler / easier to follow if we limit filteredFns to `nPrint` candidates here. You could rename the variable `printFns`. I wouldn't expect this to require you to change any `.good` files. What you have is OK if you don't want to pursue this.

Reply:
To make sure I'm understanding, the idea is:

* once we reach this line, `filteredFns` will always have `nPrint` or fewer candidates because that's all we'll add to it
* then the logic below gets simplified because we never hit the `i > nPrint` condition and don't have to worry about the `i <= nPrint` condition.  In essence, we print "all" the filtered functions and then have a simpler condition for printing out the "other candidates" addendum

Did I get that right?  (I found that logic pretty squirrelly, so like the idea of simplifying it)

Reply:
Took a stab at this in https://github.com/chapel-lang/chapel/pull/16969/commits/a60e22fc990a0fda59e97b7e005fb23a09eb6e00.  Running a full test now.

Reply:
Looks like I didn't get something quite right.  Will take another look tomorrow.

Reply:
Yes that is what I was proposing.